### PR TITLE
Default to `jnp.float_` type in `nn.initializers`

### PR DESCRIPTION
**Pull request by romanngg**
Currently initializers default to `jnp.float32`, which is not consistent with defaults in `jax.random`, which uses `jnp.float_`.

### CODE REVIEW

**Comment by jakevdp** [comment]
Thanks! To fix the test issues, I think the best approach would be to explicitly call `dtypes.canonicalize_dtype()` on each input. I think this is also the approach used in `jax.random`

**Comment by romanngg** [comment]
Thanks Jake! Added your suggestion, now most tests pass, except for `numpy-dispatch`, do you know what might cause this specific setting to fail?

**Comment by jakevdp** [comment]
Hmm... I think this is an X64 thing rather than a numpy-dispatch thing. It looks like stax somewhere assumes that the default type here will be float32, even in X64 mode.

**Comment by romanngg** [comment]
Thanks Jake! Fixed the failing tests.

File: tests/xmap_test.py [hunk 1]
```diff
@@ -734,7 +734,7 @@ def testVarianceScaling(self, map_in, map_out, fan, distr):
     shape = (80, 50, 7)
     fan_in, fan_out = jax._src.nn.initializers._compute_fans(
         NamedShape(*shape), 0, 1)
-    key = jax.random.PRNGKey(0)
+    key = jax.random.PRNGKey(1)
```
Comment (jakevdp) [inline tests/xmap_test.py:737]: Why this change of seed?

Comment (romanngg) [reply tests/xmap_test.py:737]: The test was failing otherwise after the change (from x32 params to x64), I changed the seed to make sure it's random and not resulting it worse precision. I can up the tolerance alternatively, lmk!

Comment (jakevdp) [reply tests/xmap_test.py:737]: No, that's fine - just curious

**Review by jakevdp — APPROVED** [review]
Looks good! I'm going to run this through an extra round of tests because changing the default dtype can sometimes break downstream code in unexpected ways.
